File: ArchiveWatch/ArchiveWatch/Services/PlaylistShare.swift
import Compression
import Foundation

/// A playlist you can hand to somebody, as a link.
///
/// THE PLAYLIST IS THE LINK. This project has no backend and no accounts
/// (Decisions 009 and 028), and `privacy.html` promises a playlist never
/// leaves the device except through the viewer's own cloud. A sharing service
/// would undo both, and measurement says one is unnecessary: archive ids are
/// short (median 23 characters), so deflated and base64url'd a 50-title
/// playlist is about 1,200 characters — inside what messages, mail and QR
/// codes carry without mangling.
///
/// THE FORMAT IS THE WEB'S. `js/watch.js`'s `ShareList` is the reference
/// implementation; this must produce blobs it can read and read blobs it
/// produces, because a link made on an Apple TV is opened in someone else's
/// browser. Raw DEFLATE is the codec precisely because every platform here
/// already has it: `COMPRESSION_ZLIB` is raw deflate (see
/// `CatalogRefreshService`, which inflates the catalogue with it) and Android
/// has `Deflater(level, nowrap: true)`.
///
/// A `0` PREFIX MEANS UNCOMPRESSED, and it exists for Roku. BrightScript has
/// no deflate — it has base64 on `roByteArray` and nothing else — so a
/// platform that cannot compress may emit the same JSON uncompressed behind
/// that marker rather than being locked out of sharing. Every decoder accepts
/// both; only encoders choose.
enum PlaylistShare {

    /// Titles past this make a link that messaging apps and QR codes start to
    /// mangle. Measured, not guessed: 50 real ids encode to ~1,223 characters.
    static let limit = 50

    static func blob(name: String, archiveIDs: [String]) -> String? {
        let payload: [String: Any] = ["n": String(name.prefix(80)), "i": archiveIDs]
        guard let json = try? JSONSerialization.data(withJSONObject: payload,
                                                     options: [.withoutEscapingSlashes])
        else { return nil }
        guard let deflated = deflate(json) else {
            // Uncompressed fallback: a shareable link beats no link at all.
            return "0" + base64url(json)
        }
        return base64url(deflated)
    }

    static func url(name: String, archiveIDs: [String]) -> URL? {
        guard !archiveIDs.isEmpty, archiveIDs.count <= limit,
              let b = blob(name: name, archiveIDs: archiveIDs) else { return nil }
        // The PATH is `/list/` and the playlist rides the FRAGMENT. A browser
        // never sends a fragment to a server, so the list stays off ours —
        // while the path is what a native app can match, because an Android
        // intent filter matches the path and cannot see a fragment at all.
        // `/list/` is also a real 200 page: several crawlers decline to
        // preview a 404, and this link is made to be posted.
        return URL(string: "https://archivewatch.org/list/#\(b)")
    }

    // MARK: - reading a link somebody sent

    /// What a shared link carries.
    struct Shared: Equatable {
        let name: String
        let archiveIDs: [String]
    }

    /// Pull the blob out of a share link, in either shape it has ever had.
    ///
    /// `/list/#<blob>` is what every platform emits now; `#/list/<blob>` is
    /// what shipped first and is already in the wild. Links are permanent, so
    /// both are read forever — only encoders ever choose (the same rule the
    /// `0` prefix follows).
    static func blob(from url: URL) -> String? {
        let s = url.absoluteString
        if let r = s.range(of: "/list/#") { return trimmed(String(s[r.upperBound...])) }
        if let r = s.range(of: "#/list/") { return trimmed(String(s[r.upperBound...])) }
        // A blob written into the path rather than the fragment. Nothing emits
        // this, and 404.html accepts it, so the app does too.
        let parts = url.pathComponents.filter { $0 != "/" }
        if let i = parts.firstIndex(of: "list"), i + 1 < parts.count { return trimmed(parts[i + 1]) }
        return nil
    }

    private static func trimmed(_ s: String) -> String? {
        let v = s.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return v.isEmpty ? nil : v
    }

    /// Decode a blob into the playlist it carries. Accepts both variants: a
    /// leading `0` is the uncompressed one Roku emits, anything else is raw
    /// DEFLATE.
    static func decode(_ blob: String) -> Shared? {
        let bytes: Data?
        if blob.hasPrefix("0") {
            bytes = data(base64url: String(blob.dropFirst()))
        } else {
            guard let d = data(base64url: blob) else { return nil }
            bytes = inflate(d)
        }
        guard let json = bytes,
              let o = try? JSONSerialization.jsonObject(with: json) as? [String: Any],
              let ids = o["i"] as? [String]
        else { return nil }
        // The name is optional on the wire; a playlist without one is still a
        // playlist, and the web's reference decoder says so too.
        let name = (o["n"] as? String) ?? "Shared playlist"
        return Shared(name: name, archiveIDs: ids)
    }

    static func shared(from url: URL) -> Shared? {
        guard let b = blob(from: url) else { return nil }
        return decode(b)
    }

    // MARK: - the two primitives

    /// Raw DEFLATE, the same shape `deflate-raw` produces in the browser.
    private static func deflate(_ data: Data) -> Data? {
        guard !data.isEmpty else { return nil }
        // Deflate can expand incompressible input; the margin is generous
        // because a wrong-sized buffer here is a silent truncation.
        let cap = data.count + 1_024
        let dst = UnsafeMutablePointer<UInt8>.allocate(capacity: cap)
        defer { dst.deallocate() }
        let n = data.withUnsafeBytes { src -> Int in
            guard let base = src.bindMemory(to: UInt8.self).baseAddress else { return 0 }
            return compression_encode_buffer(dst, cap, base, data.count, nil, COMPRESSION_ZLIB)
        }
        return n > 0 ? Data(bytes: dst, count: n) : nil
    }

    /// Raw INFLATE. `COMPRESSION_ZLIB` is raw deflate on Apple platforms, the
    /// same codec `CatalogRefreshService` uses for the catalogue.
    private static func inflate(_ data: Data) -> Data? {
        guard !data.isEmpty else { return nil }
        // A playlist is small and bounded by `limit`, so one generous buffer
        // beats a streaming decode. 64 KB is far past a 50-title payload.
        let cap = 64 * 1_024
        let dst = UnsafeMutablePointer<UInt8>.allocate(capacity: cap)
        defer { dst.deallocate() }
        let n = data.withUnsafeBytes { src -> Int in
            guard let base = src.bindMemory(to: UInt8.self).baseAddress else { return 0 }
            return compression_decode_buffer(dst, cap, base, data.count, nil, COMPRESSION_ZLIB)
        }
        return n > 0 ? Data(bytes: dst, count: n) : nil
    }

    /// base64url -> bytes. The padding was stripped on the way out, so it has
    /// to be put back: Foundation refuses an unpadded string.
    private static func data(base64url s: String) -> Data? {
        var t = s.replacingOccurrences(of: "-", with: "+")
                 .replacingOccurrences(of: "_", with: "/")
        let pad = (4 - t.count % 4) % 4
        t += String(repeating: "=", count: pad)
        return Data(base64Encoded: t)
    }

    /// base64url with the padding stripped — what survives a URL fragment.
    private static func base64url(_ d: Data) -> String {
        d.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}

File: android/app/src/main/java/app/archivewatch/android/data/PlaylistShare.kt
package app.archivewatch.android.data

import android.util.Base64
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import android.net.Uri
import java.util.zip.Deflater
import java.util.zip.Inflater

/**
 * A playlist you can hand to somebody, as a link.
 *
 * THE PLAYLIST IS THE LINK. There is no backend and no account here
 * (Decisions 009 and 028), and privacy.html promises a playlist never leaves
 * the device except through the viewer's own cloud. Measurement says no
 * service is needed: archive ids are short (median 23 characters), so
 * deflated and base64url'd a 50-title playlist is about 1,200 characters —
 * inside what messaging apps and QR codes carry without mangling.
 *
 * THE FORMAT IS THE WEB'S, and it has to be exactly: a link made on a Google
 * TV is opened in somebody else's browser. `js/watch.js`'s `ShareList` is the
 * reference. Raw DEFLATE is the codec because every platform already has it —
 * here that is `Deflater(level, nowrap = true)`, which is the same bytes
 * `deflate-raw` produces in a browser and `COMPRESSION_ZLIB` produces on
 * Apple.
 */
object PlaylistShare {

    /** Measured, not guessed: 50 real ids encode to ~1,223 characters. */
    const val LIMIT = 50

    fun blob(name: String, archiveIDs: List<String>): String {
        val json = JSONObject()
            .put("n", name.take(80))
            .put("i", JSONArray(archiveIDs))
            .toString()
            .toByteArray(Charsets.UTF_8)
        val deflated = deflate(json)
        // A '0' prefix means uncompressed — the variant that lets a platform
        // without deflate (Roku) share too. Everything that CAN compress does.
        return if (deflated != null) b64(deflated) else "0" + b64(json)
    }

    fun url(name: String, archiveIDs: List<String>): String? {
        if (archiveIDs.isEmpty() || archiveIDs.size > LIMIT) return null
        // the PATH is /list/ and the playlist rides the FRAGMENT: a browser never
        // sends a fragment to a server, so the list stays off ours, while the path
        // is what a native app can match (an Android intent filter cannot see a
        // fragment at all).
        // /list/ is also a real 200 page, and several crawlers decline to preview
        // a 404 — which matters for a link made to be posted.
        return "https://archivewatch.org/list/#" + blob(name, archiveIDs)
    }

    /** What a shared link carries. */
    data class Shared(val name: String, val archiveIDs: List<String>)

    /**
     * Pull the blob out of a share link, in either shape it has ever had.
     *
     * `/list/#<blob>` is what every platform emits now; `#/list/<blob>` shipped
     * first and is already in the wild. Links are permanent, so both are read
     * forever — only encoders ever choose, the same rule the '0' prefix follows.
     *
     * NOTE the fragment. An Android intent filter matches the PATH and cannot
     * see a fragment at all, which is exactly why the path carries `/list/` and
     * the payload rides behind the `#`: the filter can match, and the playlist
     * still never reaches a server.
     */
    fun blobFrom(uri: Uri): String? {
        val s = uri.toString()
        s.indexOf("/list/#").let { if (it >= 0) return s.substring(it + 7).trim('/').ifEmpty { null } }
        s.indexOf("#/list/").let { if (it >= 0) return s.substring(it + 7).trim('/').ifEmpty { null } }
        // A blob written into the path instead. Nothing emits this; 404.html
        // accepts it, so this does too.
        val segs = uri.pathSegments
        val i = segs.indexOf("list")
        if (i >= 0 && i + 1 < segs.size) return segs[i + 1].ifEmpty { null }
        return null
    }

    /**
     * Decode a blob into the playlist it carries. A leading '0' is the
     * uncompressed variant Roku emits; anything else is raw DEFLATE.
     */
    fun decode(blob: String): Shared? = runCatching {
        // A shared link that will not open is a dead end for whoever was sent
        // it, so this says WHY rather than returning a silent null (the
        // project's own debugging rule). One line, only on failure.
        Log.d("AWSHARE", "decode len=" + blob.length + " prefix=" + blob.take(8))
        val bytes = if (blob.startsWith("0")) {
            Base64.decode(blob.substring(1), Base64.URL_SAFE)
        } else {
            inflate(Base64.decode(blob, Base64.URL_SAFE)) ?: return null
        }
        val o = JSONObject(String(bytes, Charsets.UTF_8))
        val arr = o.optJSONArray("i") ?: return null
        val ids = (0 until arr.length()).map { arr.getString(it) }
        // The name is optional on the wire; a playlist without one is still a
        // playlist, and the web's reference decoder says so too.
        Shared(o.optString("n", "Shared playlist").ifEmpty { "Shared playlist" }, ids)
    }.onFailure { Log.w("AWSHARE", "decode failed: " + it) }.getOrNull()

    fun sharedFrom(uri: Uri): Shared? {
        val b = blobFrom(uri)
        if (b == null) { Log.d("AWSHARE", "no blob in " + uri); return null }
        return decode(b)
    }

    /** Raw INFLATE, the inverse of the deflate below — `nowrap = true` again. */
    private fun inflate(data: ByteArray): ByteArray? = runCatching {
        val inf = Inflater(true)
        inf.setInput(data)
        // A playlist is small and bounded by LIMIT, so one generous buffer
        // beats a streaming decode. 64 KB is far past a 50-title payload.
        val out = ByteArray(64 * 1024)
        val n = inf.inflate(out)
        inf.end()
        if (n > 0) out.copyOf(n) else null
    }.getOrNull()

    /** Raw DEFLATE — `nowrap = true` is what drops the zlib header. */
    private fun deflate(data: ByteArray): ByteArray? = runCatching {
        val d = Deflater(Deflater.BEST_COMPRESSION, true)
        d.setInput(data); d.finish()
        // Deflate can EXPAND incompressible input; a tight buffer here would
        // truncate silently, which is the worst possible failure for a link.
        val out = ByteArray(data.size + 1024)
        val n = d.deflate(out)
        d.end()
        if (n > 0) out.copyOf(n) else null
    }.getOrNull()

    private fun b64(b: ByteArray): String =
        Base64.encodeToString(b, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
}
